File: native/ios/Sources/PatientSettingsAPI.swift
import Foundation

@MainActor protocol PatientSettingsService {
    func load(identity: Identity) async throws -> PatientSettingsSnapshot
    func save(identity: Identity, snapshot: PatientSettingsSnapshot, draft: PatientSettingsDraft) async throws
}

@MainActor final class PatientSettingsAPI: PatientSettingsService {
    private let transport: NativeAPI
    private let contract: MobileContract
    private let fetch: () async throws -> SessionAccess
    init(transport: NativeAPI, contract: MobileContract, fetch: @escaping () async throws -> SessionAccess) {
        self.transport = transport; self.contract = contract; self.fetch = fetch
    }
    private func route(_ identity: Identity) throws -> ProfileContract {
        guard identity.role == .patient, let route = contract.profiles[Role.patient.rawValue] else { throw MobileFailure.forbidden }
        return route
    }
    private func access(_ identity: Identity) async throws -> SessionAccess {
        let result = try await fetch()
        try Task.checkCancellation()
        guard result.identity.subject == identity.subject, result.identity.role == identity.role else { throw MobileFailure.session }
        return result
    }
    func load(identity: Identity) async throws -> PatientSettingsSnapshot {
        let route = try route(identity)
        let session = try await access(identity)
        let path = route.readPath + (route.appendSubject ? "/" + identity.subject : "")
        let data = try await transport.request(access: session, service: route.service, path: path)
        try Task.checkCancellation()
        return try PatientSettingsData.decode(data, identity: identity, subjectField: route.subjectField)
    }
    func save(identity: Identity, snapshot: PatientSettingsSnapshot, draft: PatientSettingsDraft) async throws {
        let route = try route(identity)
        guard snapshot.subject == identity.subject else { throw MobileFailure.forbidden }
        let changes = try PatientSettingsData.changes(snapshot: snapshot, draft: draft, maxNameLength: contract.patientSettings.maxNameLength)
        guard !changes.isEmpty else { return }
        let session = try await access(identity)
        _ = try await transport.request(access: session, service: route.service,
            path: contract.patientSettings.updatePath + "/" + identity.subject, body: changes, method: .put)
        try Task.checkCancellation()
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/PatientSettingsApi.kt
package com.mediconnect.nativeapp

import org.json.JSONObject

interface PatientSettingsService {
    suspend fun load(identity: Identity): PatientSettingsSnapshot
    suspend fun save(identity: Identity, snapshot: PatientSettingsSnapshot, draft: PatientSettingsDraft)
}

class PatientSettingsApi(private val transport: NativeApi, private val contract: MobileContract) : PatientSettingsService {
    private fun route(identity: Identity): ProfileContract {
        require(identity.role == Role.PATIENT)
        return contract.profiles[Role.PATIENT] ?: throw ApiFailure(403)
    }
    override suspend fun load(identity: Identity): PatientSettingsSnapshot {
        val route = route(identity)
        val path = route.readPath + if (route.appendSubject) "/${identity.subject}" else ""
        return PatientSettingsData.decode(JSONObject(transport.request(identity, route.service, path)), identity, route.subjectField)
    }
    override suspend fun save(identity: Identity, snapshot: PatientSettingsSnapshot, draft: PatientSettingsDraft) {
        val route = route(identity)
        require(snapshot.subject == identity.subject)
        val changes = PatientSettingsData.changes(snapshot, draft, contract.patientSettings.maxNameLength)
        if (changes.length() == 0) return
        transport.request(identity, route.service, contract.patientSettings.updatePath + "/" + identity.subject,
            body = changes, method = NativeHttpMethod.PUT)
    }
}
